src/runtime/queue.h

```diff
@@ -0,0 +1,173 @@
+#include <runtime.h>
+#include <x86_64.h>
+
+/* TODO:
+   - queue paranoia mode
+   - fix kernel headers in general (shouldn't need arch-specific include above)
+*/
+
+typedef struct queue {
+    volatile union {
+        struct {
+            u32 prod_head;
+            u32 cons_tail;
+        };
+        u64 prod_combined;
+    };
+    volatile union {
+        struct {
+            u32 cons_head;
+            u32 prod_tail;
+        };
+        u64 cons_combined;
+    };
+    void ** d;
+    heap h;
+    int order;
+} *queue;
+
+#define _queue_size(q) (1ul << (q)->order)
+#define _queue_idx(q, i) ((i) & MASK((q)->order))
+#define _queue_assert(x) assert(x)
+#define _queue_pause kern_pause
+
+/* little endian */
+#define _head_from_combined(c) ((u32)((c) & MASK(32)))
```

Comment:
it doesn't matter..but doesn't the union take care of this?

Reply:
Taking from the union defeats the 64-bit copy but, yes, it makes more sense to overlay the struct onto the copy and extract from that (endian safe).